docs/cudf/source/developer_guide/ARCHITECTURE.md

```diff
@@ -0,0 +1,199 @@
+  which in turn makes shared intialization across inheritance trees much more cumbersome to manage.
+  To reenable sharing constructor logic across different index classes,
+  we instead define `BaseIndex` as the parent class of all indexes.
+  `Index` inherits from `BaseIndex`, but it masquerades as a `BaseIndex` to match pandas.
+  This class should contain no implementations since it is simply a factory for other indexes.
+
+
+## The Column layer
+
+The next layer in the cuDF stack is the Column layer.
+This layer forms the glue between pandas-like APIs and our underlying data layouts.
+Under the hood, cuDF is built around the [Apache Arrow Format](https://arrow.apache.org).
+This data format is both conducive to high-performance algorithms and suitable for data interchange between libraries.
+
+The principal objects in the Column layer are the `ColumnAccessor` and the various `Column` classes.
+The `Column` is cuDF's core data structure and is modeled after the
+[Apache Arrow Columnar Format](https://arrow.apache.org/docs/format/Columnar.html).
+A `ColumnAccessor` is a dictionary-like interface to a sequence of `Columns`.
+A `Frame` owns a `ColumnAccessor`, and most of its operations are implemented as loops over that object's `Column`s.
+
+### ColumnAccessor
+
+The primary purpose of the `ColumnAccessor` is to encapsulate pandas column selection semantics.
+Columns may be selected or inserted by index or by label, and label-based selections are as flexible as pandas is.
+For instance, Columns may be selected hierarchically (using tuples) or via wildcards.
+`ColumnAccessors` also support the `MultiIndex` columns that can result from operations like groupbys.
+
+### Columns
+
+The parent `Column` class is implemented in Cython to support interchange with C++ (more on that later).
+However, the bulk of the `Column`'s functionality is embodied by the `ColumnBase` subclass.
+`ColumnBase` provides many standard methods, while others only make sense for data of a specific type.
+As a result, we have various subclasses of `ColumnBase` like `NumericalColumn`, `StringColumn`, and `DatetimeColumn`.
+Most dtype-specific decisions should be handled at the level of a specific `Column` subclass.
+Each type of `Column` only implements methods supported by that data type.
+
+A `Column` is composed of the following:
+
+- A **data type**, specifying the type of each element.
+- A **data buffer** that may store the data for the column elements.
+  Some column types do not have a data buffer, instead storing data in the children columns.
+- A **mask buffer** whose bits represent the validity (null or not null) of each element.
+  Nullability is a core concept in the Arrow data model.
+  Columns whose elements are all valid may not have a mask buffer.
+  Mask buffers are padded to 64 bytes.
+- A tuple of **children** columns used to represent complex types with non-fixed width elements like strings or lists.
```

Comment:
Structs also use child columns, but are an exception to the rule of "non-fixed width elements" stated here.
```suggestion
- A tuple of **children** columns used to represent complex types such as non-fixed width elements (strings or lists) or structs containing nested columns.
```

Reply:
I only intended for this sentence to indicate some examples, not provide too much information as to why. That's the province of the Arrow spec. I've moved one link to the Arrow spec to just below this list to call greater attention to it since the definition of the column is probably the most Arrow-specific part of our library.